### PR TITLE
JSUI-2934 Add basePathShouldFilterResults option

### PR DESCRIPTION
**Pull request by ThibodeauJF**
WIll be part of the ongoing April release
https://coveord.atlassian.net/browse/JSUI-2934

[Deploy](https://dashboard.heroku.com/pipelines/a3535101-5bbf-4a5b-a909-47fcf8c9f149)

### CODE REVIEW

File: src/ui/DynamicHierarchicalFacet/DynamicHierarchicalFacet.ts [hunk 1]
```diff
@@ -254,7 +254,27 @@ export class DynamicHierarchicalFacet extends Component implements IDynamicHiera
...
      * @examples electronics, electronics\,laptops
      */
-    basePath: ComponentOptions.buildListOption<string>({ defaultValue: [] }),
+    basePath: ComponentOptions.buildListOption<string>({ defaultValue: [], section: 'CommonOptions' }),
+
+    /**
+     * Whether to use the [`basePath`]{@link DynamicHierarchicalFacet.options.basePath} option as a filter for the results.
+     *
+     * **Note:**
+     * Setting this option to `true` and using the [`dependsOn`]{@link DynamicHierarchicalFacet.options.dependsOn} option can cause unexpected behaviour.
```
Comment (olamothe) [inline src/ui/DynamicHierarchicalFacet/DynamicHierarchicalFacet.ts:263]: What unexpected behaviour ? Can we explain exactly what will happen ? If I'm a end user and read this, I'm not too sure what to expect, honestly...

The UI will force the option to`false`. This should be explained, and why (it's fine to say that it is un-supported).

File: src/rest/Facet/FacetRequest.ts [hunk 2]
```diff
@@ -200,6 +200,13 @@ export interface IFacetRequest {
    */
   basePath?: string[];
 
+  /**
+   * Whether to use `basePath` as a filter for the results.
+   *
+   * **Default (Search API):** `true`
```
Comment (olamothe) [inline src/rest/Facet/FacetRequest.ts:206]: I think we should explain a bit more here. I'd rephrase that, by default, the option is `undefined` and the Coveo platform will decide what will happen. And then if we can explain that the Search API will filter by default on the given facet field/base path.

File: src/ui/DynamicHierarchicalFacet/DynamicHierarchicalFacet.ts [hunk 3]
```diff
@@ -254,7 +254,27 @@ export class DynamicHierarchicalFacet extends Component implements IDynamicHiera
...
+     */
+    basePathShouldFilterResults: ComponentOptions.buildBooleanOption({
+      defaultValue: false,
```
Comment (samisayegh) [inline src/ui/DynamicHierarchicalFacet/DynamicHierarchicalFacet.ts:266]: Should the default value be `true` to match the search api default documented on line 206 of `FacetRequest.ts`?

File: src/ui/DynamicHierarchicalFacet/DynamicHierarchicalFacet.ts [hunk 4]
```diff
@@ -254,7 +254,27 @@ export class DynamicHierarchicalFacet extends Component implements IDynamicHiera
...
+      depend: 'basePath',
+      section: 'CommonOptions',
+      postProcessing: (value, options: IDynamicHierarchicalFacetOptions) => {
+        if (value && options.dependsOn) {
+          new Logger(options).warn('basePathShouldFilterResults is not compatible with dependsOn, set to false');
```
Comment (samisayegh) [inline src/ui/DynamicHierarchicalFacet/DynamicHierarchicalFacet.ts:271]: Hmm, I'm not clear on what the incompatibility is between the `basePath` and `dependsOn` options?

File: src/rest/Facet/FacetRequest.ts (hunk 2 above)
Comment (ThibodeauJF) [reply src/rest/Facet/FacetRequest.ts:206]: @fbeaudoincoveo copied the doc from the PR on SEARCH-API
Maybe it should be more descriptive?
That file is only the interface for the API request, not the JSUI option that uses that parameter.

File: src/ui/DynamicHierarchicalFacet/DynamicHierarchicalFacet.ts (hunk 3 above)
Comment (ThibodeauJF) [reply src/ui/DynamicHierarchicalFacet/DynamicHierarchicalFacet.ts:266]: Because it made more sense to have it false by default, even if for the API the default is true. The filtering part is an extra that wasn't part of the CategoryFacet.
I think if we see it from a support perspective, we will get less cases/questions if we don't filter by default.

Comment (samisayegh) [reply src/ui/DynamicHierarchicalFacet/DynamicHierarchicalFacet.ts:266]: Hmm, I'm wondering why the searchapi chose `true`, since theoretically someone could use the api directly? Looking at @olamothe's comment above:

> the option is undefined and the Coveo platform will decide what will happen

It sounds like the behaviour change will take place on older JSUI versions not containing this PR. Could the support cases come anyway? Also, if people on existing versions upgrade to the new one, won't that be a behaviour change that could create confusion?


Personally, I feel it's more natural that a user would want to see results filtered on the base path as the default behaviour rather than the reverse. Aside from support cases, what other other reasons to deactivate by default?

Comment (ThibodeauJF) [reply src/ui/DynamicHierarchicalFacet/DynamicHierarchicalFacet.ts:266]: BasePath is new to this version and the older CategoryFacet component's basePath doesn't filter results.
Might also have found a bug with selected facet values not being affected by the basePath filter correctly. if thats the case, I will remove the option temporarily.
